### PR TITLE
feat: reset last message #172 - Enable users to delete or reset last conversion between users and chatgpt, but can't delete or reset an image command message or a transcribe command message.

### PR DESCRIPTION
**Pull request by samdoghor**
Enable users to delete or reset last conversion between users and chatgpt, but can't delete or reset an image command message or a transcribe command message.

### CODE REVIEW

**Comment by samdoghor** [comment]
@n3d1117

**Comment by n3d1117** [comment]
Sorry @samdoghor, I've been busy. I'll review this PR when I have some free time

**Comment by samdoghor** [comment]
> Sorry @samdoghor, I've been busy. I'll review this PR when I have some free time

No problem.

File: bot/telegram_bot.py [hunk 1]
```diff
@@ -183,10 +199,81 @@ async def reset(self, update: Update, context: ContextTypes.DEFAULT_TYPE):
...
+
             message_thread_id=get_thread_id(update),
             text=localized_text('reset_done', self.config['bot_language'])
         )
 
+    async def unsend(self, update: Update, context: CallbackContext):
+        """
+        Unsend the last message sent by the bot
+        """
+        if not await self.is_allowed(update, context):
```
Comment (n3d1117) [inline bot/telegram_bot.py:211]: should be `if not is_allowed(update, context)`, as `is_allowed` is imported and no longer part of this class

File: bot/telegram_bot.py [hunk 2]
```diff
@@ -36,36 +36,47 @@ def __init__(self, config: dict, openai: OpenAIHelper):
         self.openai = openai
         bot_language = self.config['bot_language']
         self.commands = [
-            BotCommand(command='help', description=localized_text('help_description', bot_language)),
-            BotCommand(command='reset', description=localized_text('reset_description', bot_language)),
-            BotCommand(command='image', description=localized_text('image_description', bot_language)),
-            BotCommand(command='stats', description=localized_text('stats_description', bot_language)),
-            BotCommand(command='resend', description=localized_text('resend_description', bot_language))
+            BotCommand(command='help', description=localized_text(
+                'help_description', bot_language)),
```
Comment (n3d1117) [inline bot/telegram_bot.py:40]: There are lots of changes like this adding unneeded newlines. Why is that? These are not required

File: bot/telegram_bot.py [hunk 3]
```diff
@@ -183,10 +199,81 @@ async def reset(self, update: Update, context: ContextTypes.DEFAULT_TYPE):
...
+
+            # delete the user's last message
+            await context.bot.delete_message(chat_id=chat_id, message_id=third_to_last_message)
+            # delete the bot's last message
+            await context.bot.delete_message(chat_id=chat_id, message_id=second_to_last_message)
+            # await context.bot.delete_message(chat_id=chat_id, message_id=last_message) # delete the unsend command
+
+            await context.bot.send_message(chat_id=chat_id, text=localized_text('reset_done', self.config['bot_language']))
+        else:
+            await context.bot.send_message(chat_id=chat_id, text="!Opps, sorry you can't delete your last message because it not part of the main conversation")
```
Comment (n3d1117) [inline bot/telegram_bot.py:275]: This message should support various translations

File: bot/telegram_bot.py [hunk 4]
```diff
@@ -198,11 +285,25 @@ async def image(self, update: Update, context: ContextTypes.DEFAULT_TYPE):
...
                 text=localized_text('image_no_prompt', self.config['bot_language'])
             )
             return
 
+        """
+        Store this bot command id
+        """
+        if update.message.text.startswith("/image"):
+
+            if chat_id not in self.bot_commands:
```
Comment (n3d1117) [inline bot/telegram_bot.py:299]: `chat_id` is not in the scope here. You'll need to declare it somewhere before that, i.e:
```
chat_id = update.effective_chat.id
```

**Review by n3d1117 — CHANGES_REQUESTED** [review]

**Comment by n3d1117** [comment]
Hey @samdoghor thanks for your contribution! 🎉 
I've reviewed the PR and there are a few changes required before merging. Please also make sure to remove all the uneeded newlines that were added

**Comment by samdoghor** [comment]
Noted

File: bot/telegram_bot.py (hunk 2 above)
Comment (samdoghor) [reply bot/telegram_bot.py:40]: Made the PR long time ago.

**Pull request closed without merging**